service/storage_proxy.cc

```diff
@@ -4593,6 +4595,36 @@ future<bool> storage_proxy::cas(schema_ptr schema, shared_ptr<cas_request> reque
     co_return condition_met;
 }
 
+bool storage_proxy::is_likely_to_respond_in_time(const gms::inet_address& ep, clock_type::time_point timeout) {
+    if (fbu::is_me(ep)) {
+        return true;
+    }
+    if (auto last_seen_ep = last_seen(ep)) {
+        auto now = clock_type::now();
+        auto time_without_response = last_seen_ep->last_sent - last_seen_ep->last_responded;
+        auto time_left = timeout - now;
+        auto last_attempt = now - last_seen_ep->last_sent;
+        // If the node has never responded yet or there was no communication attempt for a long while,
+        // the node should be probed in case it becomes responsive
+        if (last_seen_ep->last_responded == clock_type::time_point::min() || last_attempt > time_left) {
+            return true;
+        }
+        if (time_without_response > time_left) {
+            static constexpr int granularity = 10'000;
+            static thread_local std::uniform_int_distribution dice(0, granularity);
+            // There's always a small chance of probing the replica in case it came back to life.
+            int fail_chance = std::min<long>((time_without_response.count() - time_left.count()) * granularity / time_left.count(), granularity - 1);
+            int roll = dice(_urandom);
+            if (roll < fail_chance) {
+                slogger.debug("Overload protection: speculatively disqualifying {} as a candidate for a read request with {}ms left until timeout, "
+                        "since it hasn't responded for at least {}ms", ep, time_left.count(), time_without_response.count());
+                return false;
+            }
+        }
+    }
```

Comment:
I like the probabilitic aspect, it gives an overloaded node opportunity to still function.

Reply:
I'm not convinced this mechanism can or should detect an "overloaded" node - I think rather it will detect a down or unavailable node. I tried to describe what I **think** this mechanism does and doesn't do in https://github.com/scylladb/scylla/issues/8964 - please tell me if you disagree and there is some useful sense of how this mechanism is meant to detect "overloaded" nodes.